### DISCUSSION TITLE
Automerging is inconsistent

### DISCUSSION THREAD

**Discussion by robg-temp** [renovatebot/renovate#9432] [Request Help]
As mentioned in previous questions, my team is planning to use Renovate to only update internal libraries, which has proven to be quite tricky to configure.  

The automerging seems to be a bit haphazard.  Some projects have automerge enabled, while others do not.  **Any ideas why that is?**  The automerge is configured at the root level.

![image](https://user-images.githubusercontent.com/80468671/113893703-45afc300-9795-11eb-8282-38c26517e6cd.png)

![image](https://user-images.githubusercontent.com/80468671/113893653-39c40100-9795-11eb-9524-484cf2cdeafe.png)

```
module.exports = {
  "baseBranches": ["master", "Sprint-150", "Release-152"],
  "branchPrefix": "renovate/",
  "onboarding": true,
  "onboardingConfig": {
    "extends": [ "config:base" ]
  },
  "platform": "github",
  "rangeStrategy": "bump",
  "automerge": true,
  "automergeType": "branch",
  "packageRules": [
    {
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["^@acme"],
      "enabled": false
    },
    {
      "separateMajorMinor": true,
      "major": {
        "enabled": false
      }
    },
    {
      "matchBaseBranches": ["Sprint-150", "Release-152"],
      "matchPackagePatterns": ["common-web-assets"],
      "separateMinorPatch": true,
      "minor": {
        "enabled": false
      }
    }
  ]
};
```

I tried moving automerge into the packageRules, but that had no effect:

```
module.exports = {
  "baseBranches": ["master", "Sprint-150", "Release-152"],
  "branchPrefix": "renovate/",
  "onboarding": true,
  "onboardingConfig": {
    "extends": [ "config:base" ]
  },
  "platform": "github",
  "rangeStrategy": "bump",
  "packageRules": [
    {
        "automerge": true
    },
    {
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["^@acme"],
      "enabled": false
    },
    {
      "separateMajorMinor": true,
      "major": {
        "enabled": false
      }
    },
    {
      "matchBaseBranches": ["Sprint-150", "Release-152"],
      "matchPackagePatterns": ["common-web-assets"],
      "separateMinorPatch": true,
      "minor": {
        "enabled": false
      }
    }
  ]
};
```

**Any idea what I'm doing wrong?**

Thanks!

Rob

**Comment by rarkins** [comment]
The first package rule had no match pattern and so achieves nothing. Add matchPackagePatterns: ["*"]

**Reply by robg-temp** [reply]
OK, I'll add that to my new config.

Thanks!

**Reply by robg-temp** [reply]
If I run the jenkins job with the new config, will it change existing PRs to show that automerge is now enabled?  Just want to be looking for the right signs.

**Reply by rarkins** [reply]
PR body will be updated every run

**Reply by robg-temp** [reply]
That did it.  (Except for the failing status check.  I'll be looking into that next.)

![image](https://user-images.githubusercontent.com/80468671/113914927-19ec0780-97ac-11eb-8b77-b6f933050d11.png)

Thanks again!

**Comment by robg-temp** [comment]
Based on your suggestion to add matchPackagePatterns: ["*"], I think that I will move the separateMajorMinor rule in there as well, since it does not have any matching either, and **we'd like to always exclude major updates in our matching**:
```

"packageRules": [
    {
        "matchPackagePatterns": ["*"],
        "automerge": true,
        "separateMajorMinor": true,
            "major": {
                "enabled": false
           }
    },

```

Does that look correct?

As I understand it, that rule will trickle down to the following ones. (?)

Thanks!

**Reply by rarkins** [reply]
No, you can't nest like that - packageRules should be flag. Additionally, `separateMajorMinor` already defaults to true, so there's no need to include it.

Maybe take a step back and describe in a few dot points what you're trying to achieve and then I can suggest the best way to configure it

**Reply by robg-temp** [reply]
OK, I didn't realize that.

What we're looking to do is update our own internal dependencies, i.e., those that start with "@acme".  So the rules are:

1. On **master** All packages starting with "@acme" should be updated (**minor and patch**).
2. On **release** and **sprint** branches, only the **"common"** package should be updated (**patch only**).
3. All PRs should be automerged.

So here's my latest configuration:

```
module.exports = {
  "baseBranches": ["master", "Sprint-150", "Release-152"],
  "branchPrefix": "renovate/",
  "onboarding": true,
  "onboardingConfig": {
    "extends": [ "config:base" ]
  },
  "platform": "github",
  "rangeStrategy": "bump",
  "packageRules": [
    {
       "matchPackagePatterns": ["*"],
       "automerge": true
    },
    {
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["^@acme"],
      "enabled": false
    },
    {
      "separateMajorMinor": true,
      "major": {
        "enabled": false
      }
    },
    {
      "matchBaseBranches": ["Sprint-150", "Release-152"],
      "matchPackagePatterns": ["common"],  // this is @acme/common.  Maybe use exact name instead of pattern?
      "separateMinorPatch": true,
      "minor": {
        "enabled": false
      }
    }
  ]
};
```

Thanks again!

**Reply by rarkins** [reply]
FYI it's not typical or recommended to put this amount of repo-specific config into the bot's backend config. It doesn't really scale once you get more than 1-2 repositories. Also you're going to need to keep updating your sprint/release branch names regularly.

I have used the optional `description` field below so that each rule is described for readability.

```
module.exports = {
  "baseBranches": ["master", "Sprint-150", "Release-152"],
  "onboardingConfig": {
    "extends": [ "config:base" ]
  },
  "platform": "github",
  "rangeStrategy": "bump",
  "separateMinorPatch": true,
  "packageRules": [
    {
       "description": "Disable all updates on all base branches to begin with",
       "matchPackagePatterns": ["*"],
       "enabled": false
    },
    {
      "description": "Enable @acme packages on master",
      "matchBaseBranches": ["master"],
      "matchPackagePatterns": ["^@acme"],
      "enabled": true
    },
    {
      "description": "Disable @acme major updates on master",
      "matchBaseBranches": ["master"],
      "matchPackagePatterns": ["^@acme"],
      "matchUpdateTypes": ["major"],
      "enabled": false
    },
    {
      "description": "Enable @acme/common on sprint and release branch",
      "matchBaseBranches": ["Sprint-150", "Release-152"],
      "matchPackageNames": ["@acme/common"],
      "enabled": true
    },
    {
      "description": "Disable @acme/common minor and major updates on sprint and release branch",
      "matchBaseBranches": ["Sprint-150", "Release-152"],
      "matchPackageNames": ["@acme/common"],
      "matchUpdateTypes": ["major", "minor"],
      "enabled": false
    },
    {
      "description": "automerge every PR",
      "matchPackagePatterns": ["*"],
      "automerge": true
    }
  ]
};
```

**Reply by robg-temp** [reply]
Wow, my config wasn't as close as I thought it was!  

The main reason for putting all of the rules together was mainly to get them working.  Once we add more projects, I think that we will move a lot of the rules to the individual project renovate.json files - especially automerge, which we will probably want to turn off in some cases.

I'll give your config a try today.

Thanks so much for your help with this!  Now I'm convinced that I would have been flailing around for who knows how many more weeks.

**Reply by rarkins** [reply]
There are multiple ways to do things, and I've prioritized readability over brevity, e.g. sometimes using two understandable rules instead of 1 slightly confusing one. There's usually more than one way to do things
